### PR TITLE
fix-small-bugs-on-ansible-modal

### PR DESCRIPTION
**Pull request by ChunxiAlexLuo**
Signed-off-by: Chunxi Luo chuluo@redhat.com

fix https://github.com/open-cluster-management/backlog/issues/12551

here will fix oversized link by using latest ui-component 
https://github.com/open-cluster-management/ui-components/pull/341

Screen Shot 2021-05-17 at 1 11 10 PM

### CODE REVIEW

**Comment by ChunxiAlexLuo** [comment]
fixed these two bugs first since they break update automation/ update manual mode, will also fix YAML and modal text this morning.

File: src-web/nls/platform.properties [hunk 1]
```diff
@@ -170,6 +170,7 @@ modal.ansible.scheduleMode.once.title = Run once mode
 modal.ansible.scheduleMode.once.info = When this policy is violated, the automation will run one time. Once it runs, it will be set to disabled.
 modal.ansible.scheduleMode.disabled.title = Disable automation
 modal.ansible.scheduleMode.disabled.info = Automation won't run until you change this setting.
+modal.ansible.unsaved.data = You have made the change on configuration, do you want to save your policy automation? Or re-click 'Cancel' button to close.
```
Comment (ycao56) [inline src-web/nls/platform.properties:173]: @dockerymick can you review?

Comment (dockerymick) [reply src-web/nls/platform.properties:173]: I recommend saying, "You made changes to the configuration, do you want to save your policy automation. Click the Cancel button to close"

Can you italicize Cancel or bold it instead of the single quotes?

Comment (ChunxiAlexLuo) [reply src-web/nls/platform.properties]: Sure

File: src-web/components/modals/AnsibleAutomationModal.js [hunk 2]
```diff
@@ -34,6 +34,7 @@ import {
   GET_ANSIBLE_JOB_TEMPLATE,
 } from '../../utils/client/queries'
 import _ from 'lodash'
+import {stringify} from 'flatted'
```
Comment (ycao56) [inline src-web/components/modals/AnsibleAutomationModal.js:37]: is this needed? can you use JSON.stringify?

File: src-web/components/modals/AnsibleAutomationModal.js [hunk 3]
```diff
@@ -164,18 +173,64 @@ export class AnsibleAutomationModal extends React.Component {
...
+      handleClose(modalType)
+    } else {
+      const { locale } = this.context
+      const { initialJSON, confirmClose } = this.state
+      const { latestJSON } = await this.generateJSON()
+      let ifChanged = false
+      console.log(JSON.stringify(initialJSON))
+      console.log(JSON.stringify(latestJSON))
+      if (initialJSON && latestJSON) {
+        if (stringify(initialJSON) !== stringify(latestJSON)) {
```
Comment (ycao56) [inline src-web/components/modals/AnsibleAutomationModal.js:209]: can you try using lodash isEqual to compare the object directly?
https://lodash.com/docs/4.17.15#isEqual

File: src-web/nls/platform.properties [hunk 4]
```diff
@@ -146,30 +146,34 @@ modal.enforce-hcmcompliance.heading = Enforce policy
...
-modal.ansible.scheduleMode.manual.title = Manual run
-modal.ansible.scheduleMode.manual.info = Manual make this automation immediately run once. Once it runs, it will be set to disabled.
-modal.ansible.scheduleMode.once.title = Run once mode
-modal.ansible.scheduleMode.once.info = When this policy is violated, the automation will run one time. Once it runs, it will be set to disabled.
-modal.ansible.scheduleMode.disabled.title = Disable automation
-modal.ansible.scheduleMode.disabled.info = Automation won't run until you change this setting.
+
+ansible.automation.heading.create = Create policy violation automation
+ansible.automation.heading.edit = Edit policy violation automation
+ansible.automation.description.create = Create a new provider credentials to automate your selected policy and violation. You can also schedule the automation to run once, continuously, or disable for now.
```
Comment (dockerymick) [inline src-web/nls/platform.properties:156]: I suggest: 

Create new provider credentials to automate the selected policy and violation. You can also schedule the automation to run once, continuously, or disable.

File: src-web/nls/platform.properties [hunk 5]
```diff
@@ -146,30 +146,34 @@ modal.enforce-hcmcompliance.heading = Enforce policy
...
+ansible.automation.description.edit = Modify an existing provider credentials to automate your selected policy and violation. You can also schedule the automation to run once, continuously, or disable for now.
```
Comment (dockerymick) [inline src-web/nls/platform.properties:157]: Modify existing provider credentials to automate the selected policy and violation. You can also schedule the automation to run once, continuously, or disable.

File: src-web/nls/platform.properties [hunk 6]
```diff
@@ -146,30 +146,34 @@ modal.enforce-hcmcompliance.heading = Enforce policy
...
+ansible.launch.connection = Launch to connection URL
```
Comment (dockerymick) [inline src-web/nls/platform.properties:158]: Launch to Ansible URL

File: src-web/nls/platform.properties [hunk 7]
```diff
@@ -146,30 +146,34 @@ modal.enforce-hcmcompliance.heading = Enforce policy
...
+ansible.credential.selection.title = Provider credential
+ansible.credential.selection.placeholder = Select Ansible Credential
```
Comment (dockerymick) [inline src-web/nls/platform.properties:160]: Select Ansible credential

File: src-web/nls/platform.properties [hunk 8]
```diff
@@ -146,30 +146,34 @@ modal.enforce-hcmcompliance.heading = Enforce policy
...
+ansible.no.credential = Ansible Credential is required to create Ansible Job, create your console here:
```
Comment (dockerymick) [inline src-web/nls/platform.properties:161]: Ansible credential is required to create an Ansible job. Create your console here:

Is "here" a form that the user has to fill out? If so, i would specify where the user can create the console, for example:

Create your console by completing the following form:

File: src-web/nls/platform.properties [hunk 9]
```diff
@@ -146,30 +146,34 @@ modal.enforce-hcmcompliance.heading = Enforce policy
...
+ansible.launch.createCredential = Create a new provider credentials
```
Comment (dockerymick) [inline src-web/nls/platform.properties:162]: Create new provider credentials

File: src-web/nls/platform.properties [hunk 10]
```diff
@@ -146,30 +146,34 @@ modal.enforce-hcmcompliance.heading = Enforce policy
...
+ansible.jobTemplates.selection.title = Job template
+ansible.jobTemplates.selection.placeholder = Select Ansible Job Template
```
Comment (dockerymick) [inline src-web/nls/platform.properties:164]: Select Ansible job template

File: src-web/nls/platform.properties [hunk 11]
```diff
@@ -146,30 +146,34 @@ modal.enforce-hcmcompliance.heading = Enforce policy
...
+ansible.jobTemplates.editor.title = Ansible Job Template Extra Variables
```
Comment (dockerymick) [inline src-web/nls/platform.properties:165]: in my opinion, i would reword this to read:

Extra variables in the Ansible job template

File: src-web/nls/platform.properties [hunk 12]
```diff
@@ -146,30 +146,34 @@ modal.enforce-hcmcompliance.heading = Enforce policy
...
+ansible.jobTemplates.editor.placeholder = Input Ansible Job Template Extra Variables Here
```
Comment (dockerymick) [inline src-web/nls/platform.properties:166]: I suggest the following:

Enter the extra variables from the Ansible job template

File: src-web/nls/platform.properties [hunk 13]
```diff
@@ -146,30 +146,34 @@ modal.enforce-hcmcompliance.heading = Enforce policy
...
+ansible.no.jobTemplates.info = Failed to get Ansible Job Template information, please check related Ansible Tower and Ansible Job Template setting
```
Comment (dockerymick) [inline src-web/nls/platform.properties:167]: I suggest the following:

Failed to get Ansible job template information. Check the related Ansible Tower and Ansible job template settings

File: src-web/nls/platform.properties [hunk 14]
```diff
@@ -146,30 +146,34 @@ modal.enforce-hcmcompliance.heading = Enforce policy
...
+ansible.scheduleMode.title = Schedule automation
+ansible.scheduleMode.info = Choose how often you want this automation to run. You can also choose to manually run it right now, which will save your change as is.
```
Comment (dockerymick) [inline src-web/nls/platform.properties:169]: I suggest: 

Choose how often you want this automation to run. You can also choose to manually run it, which saves your change as is.

File: src-web/nls/platform.properties [hunk 15]
```diff
@@ -146,30 +146,34 @@ modal.enforce-hcmcompliance.heading = Enforce policy
...
+ansible.scheduleMode.manual.title = Manual run
+ansible.scheduleMode.manual.info = Manual make this automation immediately run once. Once it runs, it will be set to disabled.
```
Comment (dockerymick) [inline src-web/nls/platform.properties:171]: Manually set this automation to run once. After the automation runs, it is set to disabled.

File: src-web/nls/platform.properties [hunk 16]
```diff
@@ -146,30 +146,34 @@ modal.enforce-hcmcompliance.heading = Enforce policy
...
+ansible.scheduleMode.once.title = Run once mode
+ansible.scheduleMode.once.info = When this policy is violated, the automation will run one time. Once it runs, it will be set to disabled.
```
Comment (dockerymick) [inline src-web/nls/platform.properties:173]: When this policy is violated, the automation runs one time. After the automation runs, it is set to disabled.

File: src-web/nls/platform.properties [hunk 17]
```diff
@@ -146,30 +146,34 @@ modal.enforce-hcmcompliance.heading = Enforce policy
...
+ansible.scheduleMode.disabled.title = Disable automation
+ansible.scheduleMode.disabled.info = Automation won't run until you change this setting.
```
Comment (dockerymick) [inline src-web/nls/platform.properties:175]: Automation does not run until this setting is updated.

File: src-web/nls/platform.properties [hunk 18]
```diff
@@ -146,30 +146,34 @@ modal.enforce-hcmcompliance.heading = Enforce policy
...
+ansible.unsaved.data = You have made the change on configuration, do you want to save your policy automation? Or re-click 'Cancel' button to close.
```
Comment (dockerymick) [inline src-web/nls/platform.properties:176]: I suggest:

You made changes to the configuration. Do you want to save your policy automation? Click the Cancel button to close.

Instead of the single quotes, are you able to bold the word Cancel?

File: tests/jest/config/platform-properties.json [hunk 19]
```diff
@@ -119,30 +119,33 @@
...
-  "modal.ansible.no.jobTemplates.info": "Failed to get Ansible Job Template information, please check related Ansible Tower and Ansible Job Template setting",
-  "modal.ansible.scheduleMode.title": "Schedule automation",
-  "modal.ansible.scheduleMode.info": "Choose how often you want this automation to run. You can also choose to manually run it right now, which will save your change as is.",
-  "modal.ansible.scheduleMode.manual.title": "Manual run",
-  "modal.ansible.scheduleMode.manual.info": "Manual make this automation immediately run once. Once it runs, it will be set to disabled.",
-  "modal.ansible.scheduleMode.once.title": "Run once mode",
-  "modal.ansible.scheduleMode.once.info": "When this policy is violated, the automation will run one time. Once it runs, it will be set to disabled.",
-  "modal.ansible.scheduleMode.disabled.title": "Disable automation",
-  "modal.ansible.scheduleMode.disabled.info": "Automation won't run until you change this setting.",
+  "ansible.automation.heading.create": "Create policy violation automation",
```
Comment (dockerymick) [inline tests/jest/config/platform-properties.json:126]: Please make the same updates as suggested in the previous file (`/nls/platform.properties`)

**Review by dockerymick — CHANGES_REQUESTED** [review]

File: package.json [hunk 20]
```diff
@@ -107,7 +107,8 @@
     "sass-loader": "^8.0.2",
     "source-map-loader": "^1.1.3",
     "style-loader": "^1.3.0",
-    "webpack-cli": "^4.7.0"
+    "webpack-cli": "^4.7.0",
+    "yamljs": "^0.3.0"
```
Comment (ycao56) [inline package.json:111]: this seems not used anywhere. there is already a module called `js-yaml` being imported.

Comment (ChunxiAlexLuo) [reply package.json:111]: will switch to `js-yaml`

File: src-web/nls/platform.properties [hunk 21]
```diff
@@ -146,30 +146,35 @@ modal.enforce-hcmcompliance.heading = Enforce policy
...
+ansible.automation.description.edit = Modify existing provider credentials to automate the selected policy and violation. You can also schedule the automation to run once, continuously, or disable.
+ansible.launch.connection = Launch to Ansible URL
+ansible.credential.selection.title = Credential
+ansible.credential.selection.placeholder = Select Ansible credential
+ansible.no.credential = Ansible credential is required to create an Ansible job. Create your console by clicking the following link:
+ansible.launch.createCredential = Create a new provider credentials
+ansible.jobTemplates.selection.title = Job template
+ansible.jobTemplates.selection.placeholder = Select Ansible job template
+ansible.jobTemplates.editor.title = Extra variables in the Ansible job template
+ansible.jobTemplates.editor.placeholder = Input Ansible Job Template Extra Variables Here
```
Comment (dockerymick) [inline src-web/nls/platform.properties:166]: I would change this to read: 

Enter extra variables from the Ansible job template here

File: tests/jest/config/platform-properties.json [hunk 22]
```diff
@@ -119,30 +119,34 @@
...
+  "ansible.automation.heading.edit": "Edit policy violation automation",
+  "ansible.automation.description.create": "Create new provider credentials to automate the selected policy and violation. You can also schedule the automation to run once, continuously, or disable.",
+  "ansible.automation.description.edit": "Modify existing provider credentials to automate the selected policy and violation. You can also schedule the automation to run once, continuously, or disable.",
+  "ansible.launch.connection": "Launch to Ansible URL",
+  "ansible.credential.selection.title": "Credential",
+  "ansible.credential.selection.placeholder": "Select Ansible credential",
+  "ansible.no.credential": "Ansible credential is required to create an Ansible job. Create your console by clicking the following link:",
+  "ansible.launch.createCredential": "Create a new provider credentials",
```
Comment (dockerymick) [inline tests/jest/config/platform-properties.json:134]: Create new provider credentials.

File: tests/jest/config/platform-properties.json [hunk 23]
```diff
@@ -119,30 +119,34 @@
...
+  "ansible.jobTemplates.selection.title": "Job template",
+  "ansible.jobTemplates.selection.placeholder": "Select Ansible job template",
+  "ansible.jobTemplates.editor.title": "Extra variables in the Ansible job template",
+  "ansible.jobTemplates.editor.placeholder": "Input Ansible Job Template Extra Variables Here",
```
Comment (dockerymick) [inline tests/jest/config/platform-properties.json:138]: Enter extra variables from the Ansible job template here

File: tests/jest/config/platform-properties.json [hunk 24]
```diff
@@ -119,30 +119,34 @@
...
+  "ansible.scheduleMode.title": "Schedule automation",
+  "ansible.scheduleMode.info": "Choose how often you want this automation to run. You can also choose to manually run it, which saves your change as is.",
+  "ansible.scheduleMode.manual.title": "Manual run",
+  "ansible.scheduleMode.manual.info": "Manually set this automation to run once. After the automation runs, it is set to disabled.",
+  "ansible.scheduleMode.once.title": "Run once mode",
+  "ansible.scheduleMode.once.info": "When this policy is violated, the automation runs one time. After the automation runs, it is set to disabled.",
+  "ansible.scheduleMode.disabled.title": "Disable automation",
+  "ansible.scheduleMode.disabled.info": "Automation does not run until this setting is updated.",
+  "ansible.unsaved.data": "You made changes to the configuration. Do you want to save your policy automation? Click the Cancel button to close.",
+  "ansible.loading.info": "Loading policy violation automation data...",
```
Comment (ChunxiAlexLuo) [inline tests/jest/config/platform-properties.json:149]: @dockerymick This is the loading message with spinner when first time open the panel and still loading backend data

**Review by dockerymick — APPROVED** [review]
Thanks for pulling me in @ycao and @ChunxiAlexLuo. Thanks for accepting my suggestions as well. The changes look good to me
